includes/class-indieauth-authenticate.php

```diff
@@ -0,0 +1,252 @@
+<?php
+/**
+ * Authentication class
+ * Helper functions for extracting tokens from the WP-API team Oauth2 plugin
+ */
+class IndieAuth_Authenticate {
+
+	public function __construct() {
+		add_filter( 'determine_current_user', array( $this, 'determine_current_user' ), 11 );
+		add_filter( 'rest_authentication_errors', array( $this, 'rest_authentication_errors' ) );
+		add_action( 'authenticate', array( $this, 'authenticate' ) );
+	}
+
+	/**
+	 * Report our errors, if we have any.
+	 *
+	 * Attached to the rest_authentication_errors filter. Passes through existing
+	 * errors registered on the filter.
+	 *
+	 * @param WP_Error|null Current error, or null.
+	 *
+	 * @return WP_Error|null Error if one is set, otherwise null.
+	 */
+	public function rest_authentication_errors( $error = null ) {
+		if ( ! empty( $error ) ) {
+			return $error;
+		}
+		global $indieauth_error;
```

Comment:
why make it global?

Reply:
There's no other way to surface the error. I was stumped about that till I looked at the REST API teams example Oauth plugin.